File: frontend/js/views/spousal_ira.js
// Spousal IRA — per-spouse contribution limits (with catch-up) and whether
// the couple's combined earned income covers both, via /calc/spousal-ira.
// Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const NUM = [
    ['combined_earned_income_usd', 'Combined earned income ($)', 80000],
    ['working_spouse_contribution_usd', 'Working spouse contribution ($)', 7500],
    ['nonworking_spouse_contribution_usd', 'Non-working spouse contribution ($)', 7500],
    ['base_limit_usd', 'Per-person base limit ($, 0 = default)', 0],
    ['catch_up_usd', 'Catch-up (50+) ($, 0 = default)', 0],
];
const CHECKS = [
    ['working_spouse_50plus', 'Working spouse 50+'],
    ['nonworking_spouse_50plus', 'Non-working spouse 50+'],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderSpousalIra(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.sira.h1.title">// SPOUSAL IRA</span></h1>
        <p class="muted small" data-i18n="view.sira.hint.intro">
            Normally an IRA contribution needs earned income. The spousal-IRA rule lets a
            married couple filing jointly fund an IRA for a spouse with little or no income, as
            long as their combined earned income covers both contributions. Each spouse is
            still capped at the annual limit ($7,500 for 2026, +$1,100 catch-up at 50+).
            Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.sira.h2.inputs">The couple</h2>
            <form id="sira-form" class="inline-form">
                ${NUM.map(([key, label, def]) => `
                    <label><span data-i18n="view.sira.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
                ${CHECKS.map(([key, label]) => `
                    <label><input type="checkbox" name="${key}"> <span data-i18n="view.sira.label.${key}">${label}</span></label>
                `).join('')}
            </form>
        </div>
        <div id="sira-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#sira-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of NUM) body[key] = Number(fd.get(key)) || 0;
        for (const [key] of CHECKS) body[key] = fd.get(key) != null;
        try {
            const r = await api.calcSpousalIra(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.sira.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#sira-result');
    const okCls = r.income_sufficient ? 'pos' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.sira.h2.result">Eligibility</h2>
            <div class="cards">
                <div class="card ${okCls}"><div class="label" data-i18n="view.sira.card.status">Income covers both?</div>
                    <div class="value ${okCls}">${r.income_sufficient ? t('view.sira.yes') : t('view.sira.no')}</div></div>
                <div class="card"><div class="label" data-i18n="view.sira.card.combined">Combined contribution</div>
                    <div class="value">${money(r.combined_contribution_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.sira.card.max">Max combined</div>
                    <div class="value">${money(r.max_combined_usd)}</div></div>
            </div>
            ${r.income_sufficient ? '' : `<p class="muted small neg" data-i18n="view.sira.warn.income">Combined earned income is below the contributions — total IRA contributions can't exceed earned income.</p>`}
            <table class="data-table">
                <thead><tr>
                    <th data-i18n="view.sira.col.spouse">Spouse</th>
                    <th data-i18n="view.sira.col.limit">Limit</th>
                    <th data-i18n="view.sira.col.allowed">Allowed</th>
                </tr></thead>
                <tbody>
                    <tr><td data-i18n="view.sira.row.working">Working</td><td>${money(r.working_spouse_limit_usd)}</td><td>${money(r.working_allowed_usd)}</td></tr>
                    <tr><td data-i18n="view.sira.row.nonworking">Non-working</td><td>${money(r.nonworking_spouse_limit_usd)}</td><td>${money(r.nonworking_allowed_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/spousal_ira.rs
//! Spousal IRA — contribution eligibility for a non-working spouse.
//!
//! Normally an IRA contribution requires earned income. The spousal-IRA rule
//! lets a married couple filing jointly fund an IRA for a spouse with little
//! or no income, as long as the **couple's combined earned income covers the
//! total of both spouses' contributions**. Each spouse is still capped at the
//! annual per-person limit (plus a catch-up at age 50+).
//!
//!   * per-spouse limit = base limit + (age 50+ ? catch-up : 0)
//!   * each contribution is capped at that spouse's limit
//!   * the two contributions together can't exceed combined earned income
//!
//! 2026 limits (web-verified, irs.gov): base $7,500, catch-up $1,100. Both
//! are inputs so the calc survives future indexing. Pure compute.

use serde::{Deserialize, Serialize};

/// 2026 IRA base contribution limit.
pub const DEFAULT_BASE_LIMIT_USD: f64 = 7_500.0;
/// 2026 age-50+ catch-up.
pub const DEFAULT_CATCH_UP_USD: f64 = 1_100.0;

#[derive(Debug, Clone, Deserialize)]
pub struct SpousalIraInput {
    /// The couple's combined earned income (often the working spouse's).
    pub combined_earned_income_usd: f64,
    pub working_spouse_contribution_usd: f64,
    pub nonworking_spouse_contribution_usd: f64,
    #[serde(default)]
    pub working_spouse_50plus: bool,
    #[serde(default)]
    pub nonworking_spouse_50plus: bool,
    /// Per-person base limit (defaults to the current $7,500 if 0).
    #[serde(default)]
    pub base_limit_usd: f64,
    /// Age-50+ catch-up (defaults to $1,100 if 0).
    #[serde(default)]
    pub catch_up_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct SpousalIraResult {
    pub working_spouse_limit_usd: f64,
    pub nonworking_spouse_limit_usd: f64,
    /// Working contribution after the per-person cap.
    pub working_allowed_usd: f64,
    /// Non-working contribution after the per-person cap.
    pub nonworking_allowed_usd: f64,
    /// The two capped contributions summed.
    pub combined_contribution_usd: f64,
    /// Most the couple could contribute = min(sum of limits, earned income).
    pub max_combined_usd: f64,
    /// True when earned income covers the (capped) contributions.
    pub income_sufficient: bool,
}

pub fn analyze(i: &SpousalIraInput) -> SpousalIraResult {
    let base = if i.base_limit_usd > 0.0 { i.base_limit_usd } else { DEFAULT_BASE_LIMIT_USD };
    let catch_up = if i.catch_up_usd > 0.0 { i.catch_up_usd } else { DEFAULT_CATCH_UP_USD };

    let working_limit = base + if i.working_spouse_50plus { catch_up } else { 0.0 };
    let nonworking_limit = base + if i.nonworking_spouse_50plus { catch_up } else { 0.0 };

    let working_allowed = i.working_spouse_contribution_usd.max(0.0).min(working_limit);
    let nonworking_allowed = i.nonworking_spouse_contribution_usd.max(0.0).min(nonworking_limit);
    let combined = working_allowed + nonworking_allowed;

    let income = i.combined_earned_income_usd.max(0.0);
    let max_combined = (working_limit + nonworking_limit).min(income);
    let income_sufficient = income >= combined;

    SpousalIraResult {
        working_spouse_limit_usd: working_limit,
        nonworking_spouse_limit_usd: nonworking_limit,
        working_allowed_usd: working_allowed,
        nonworking_allowed_usd: nonworking_allowed,
        combined_contribution_usd: combined,
        max_combined_usd: max_combined,
        income_sufficient,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> SpousalIraInput {
        SpousalIraInput {
            combined_earned_income_usd: 80_000.0,
            working_spouse_contribution_usd: 7_500.0,
            nonworking_spouse_contribution_usd: 7_500.0,
            working_spouse_50plus: false,
            nonworking_spouse_50plus: false,
            base_limit_usd: 0.0,
            catch_up_usd: 0.0,
        }
    }

    #[test]
    fn default_limit_under_50() {
        let r = analyze(&base());
        assert!((r.working_spouse_limit_usd - 7_500.0).abs() < 1e-9);
        assert!((r.nonworking_spouse_limit_usd - 7_500.0).abs() < 1e-9);
    }

    #[test]
    fn catch_up_only_for_50plus_spouse() {
        let r = analyze(&SpousalIraInput { working_spouse_50plus: true, ..base() });
        assert!((r.working_spouse_limit_usd - 8_600.0).abs() < 1e-9); // 7500 + 1100
        assert!((r.nonworking_spouse_limit_usd - 7_500.0).abs() < 1e-9); // unchanged
    }

    #[test]
    fn contribution_capped_at_limit() {
        let r = analyze(&SpousalIraInput { working_spouse_contribution_usd: 10_000.0, ..base() });
        assert!((r.working_allowed_usd - 7_500.0).abs() < 1e-9);
    }

    #[test]
    fn nonworking_spouse_funded_on_working_income() {
        // The whole point: nonworking spouse contributes on the couple's income.
        let r = analyze(&base());
        assert!((r.nonworking_allowed_usd - 7_500.0).abs() < 1e-9);
        assert!(r.income_sufficient); // 80k >> 15k combined
    }

    #[test]
    fn combined_is_sum_of_capped_contributions() {
        let r = analyze(&base());
        assert!((r.combined_contribution_usd - 15_000.0).abs() < 1e-9);
    }

    #[test]
    fn income_insufficient_when_below_contributions() {
        // Only $5k earned income → can't fund $15k of contributions.
        let r = analyze(&SpousalIraInput { combined_earned_income_usd: 5_000.0, ..base() });
        assert!(!r.income_sufficient);
        assert!((r.max_combined_usd - 5_000.0).abs() < 1e-9); // capped by income
    }

    #[test]
    fn max_combined_is_min_of_limits_and_income() {
        // Ample income → max = sum of limits (15k).
        let r = analyze(&base());
        assert!((r.max_combined_usd - 15_000.0).abs() < 1e-9);
    }

    #[test]
    fn both_50plus_doubles_catch_up() {
        let r = analyze(&SpousalIraInput {
            working_spouse_50plus: true,
            nonworking_spouse_50plus: true,
            ..base()
        });
        assert!((r.working_spouse_limit_usd - 8_600.0).abs() < 1e-9);
        assert!((r.nonworking_spouse_limit_usd - 8_600.0).abs() < 1e-9);
    }
}
